### PR TITLE
Add support for `$ref` URIs containing fragments in OpenAPI definitions and JSON schemas

### PR DESCRIPTION
**Pull request by zecakeh**
Based on #1745.

There is a first commit to add support for fragments in `$ref` URIs, to solve (simple) JSON Pointers. It also adds support for [Reference Objects](https://spec.openapis.org/oas/v3.1.0#reference-object) for OpenAPI request parameters and responses.

Preview: https://pr1751--matrix-spec-previews.netlify.app

### CODE REVIEW

**Comment by richvdh** [comment]
oh no :facepalm: 

This is the same thing as I did in https://github.com/matrix-org/matrix-spec/pull/1749 :(

**Comment by zecakeh** [comment]
I saw that, well it is kind of completely different though

**Comment by richvdh** [comment]
@zecakeh As I've written over at #1749, this approach seems sensible, but I'd like to land it independently of #1745. Could you rebase it without that?

**Comment by zecakeh** [comment]
Alright, I removed the commit with the OpenAPI definition change. I added a commit to also support refs with fragments in the CI scripts.

File: layouts/partials/json-schema/resolve-refs.html [hunk 1]
```diff
@@ -1,7 +1,10 @@
 {{/*
```
Comment (zecakeh) [inline layouts/partials/json-schema/resolve-refs.html:1]: Should we do like in #1749 and allow to override properties named `description` and `summary`?

File: scripts/helpers.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            ref = load_file_from_uri(full_ref_uri)
+
+            if fragment:
+                # The fragment should be a JSON Pointer
+                keys = fragment.strip('/').split('/')
+                for key in keys:
+                    ref = ref[key]
+
+            # Check that the references in *this* file are valid
+            result = resolve_references(urllib.parse.urlsplit(full_ref_uri).path, ref)
```
Comment (zecakeh) [inline scripts/helpers.py:56]: Contrary to the `resolve-refs` partial, this does not keep non-duplicate keys in the original schema, should it?

File: layouts/partials/json-schema/resolve-refs.html (hunk 1 above)
Comment (richvdh) [reply layouts/partials/json-schema/resolve-refs.html:1]: No, I don't think that's necessary. The validator will complain if anyone tries to override other properties, I think?

File: layouts/shortcodes/http-api.html [hunk 3]
```diff
@@ -21,6 +21,6 @@
 
 {{ $api_data := index .Site.Data.api .Params.spec .Params.api }}
 {{ $base_url := (index $api_data.servers 0).variables.basePath.default }}
-{{ $path := delimit (slice "api" $spec) "/" }}
+{{ $path := delimit (slice "api" $spec $api) "/" }}
```
[the remark this answers is not in the document]
Comment (richvdh) [reply layouts/shortcodes/http-api.html:24]: well, probably, yes. But we don't necessarily need to change that today.

**Review by richvdh — APPROVED** [review]
lgtm

File: layouts/partials/json-schema/resolve-refs.html (hunk 1 above)
Comment (zecakeh) [reply layouts/partials/json-schema/resolve-refs.html:1]: I doesn't, afaict.

Comment (richvdh) [reply layouts/partials/json-schema/resolve-refs.html:1]: oh. bother. Well, maybe something to think about adding into our validation scripts in future.

**Pull request merged**